Clean up timer usage

markdewing opened the pull request:
Convert all timers to use createTimer and remove uses of addTimer.

Timers are owned by TimerManagerClass, using a vector of std::unique pointers.   Deallocation is automatically handled and this addresses issue #1808.

The addRawTimer friend function was added for unit testing.   All the unit tests now create timers dynamically and ownership is transferred to the TimerManager when addRawTimer is called.

There are still more changes to make as suggested in #1808 and #1734.

ye-luo commented on src/Utilities/tests/test_timer.cpp:
```diff
@@ -20,6 +20,15 @@
+
+// The manager will take ownership of the timer.  This means the timer must
+// be allocated dynamically, not on the stack.
+void addRawTimer(TimerManagerClass &tm, NewTimer* t)
```
This function is added to serve FakeTimer.
FakeTimer is a derived class with extra functions to alter total_time and num_calls for testing.
I feel better remove this class and add testing APIs as friend function to the NewTimer class.
Then we test NewTimer directly instead of FakeTimer and addRawTimer is not needed.

markdewing replied:
Changed

ye-luo commented:
There is one question left. Should the timers belongs to the class or the class object? When the copy constructor is invoked, should it be a shallow copy of a timer or making a new timer object?
Currently the timer manage identifies timers by names when reporting. it behaves like per class because names are defined in the class. For example, the electron and ion particlesets are indistinguishable in the timing report. However, in some classes, the copy constructor requests creating new timers by the timer manager. In other classes, timers are references to timers owned by the manager, the copy constructor only does shallow copy. Maybe I'm over-concerned, keeping both ways is just fine.

markdewing commented:
I deleted the copy constructor and changed all the areas where it was being called.  There should be no need to call the copy constructor - all access to a timer should be through a pointer or a reference.

markdewing commented on src/QMCDrivers/QMCLinearOptimize.cpp:
```diff
@@ -40,6 +40,22 @@
 
 namespace qmcplusplus
 {
+enum LinearOptTimers
```
Currently all the timer details are not in the header file.   It seems preferable to keep it that way unless there is a strong reason to do otherwise.

ye-luo replied:
Could you explain me why you think the current way is preferable?
I didn't mean just move to the header but move into the class. In the current way, I cannot use a timer in a function defined in the header.

markdewing replied:
Yes, that's a good point.  More stuff in the headers means more recompilation if something changes (though the timers probably don't change much)
I'm wondering if we should drop the myTimer list and the enums.   As seen in this diff, it was an attempt to clean up existing practice.   But it adds code overhead.   Maybe we should just declare the timers directly and initialize in the constructor, as is done in the Fermion classes.

ye-luo commented on src/QMCDrivers/QMCLinearOptimize.cpp:
```diff
@@ -40,6 +40,22 @@
+  LinOpt_costFunction
+};
+
+TimerNameList_t<LinearOptTimers> LinearOptTimerNames{{LinOpt_generateSamples, "QMCLinearOptimize::GenerateSamples"},
```
Since it is used only once, we can have it right before setup_timers instead of global.

grahamlopez commented:
test this please

grahamlopez commented:
test this please

PDoakORNL commented on src/QMCDrivers/RMC/RMCUpdatePbyP.cpp:
```diff
@@ -34,6 +34,21 @@
+  RMC_energy
+};
+
+TimerNameList_t<RMCTimers> RMCTimerNames = {
```
`_t` seems uneeded here naming convention is enough to indicate this is a type.

jtkrogel commented:
@markdewing Does this affect the pattern used for timers in DensityMatrices1B.*?

markdewing commented:
@jtkrogel It might in the future.  There is no need to change it now.  That you had to change the timer style in that PR is one of the motivations that I think we should go with the simpler approach of individual timer variables.
